File: 0098_validate-binary-search-tree/solution.go
package _098_validate_binary_search_tree

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
二叉搜索数，中序遍历的结果是，从小到大排序的，如果后一个节点的值比前一个节点的值小则不是二叉搜索树，使用栈来模拟中序遍历过程
*/
func isValidBST(root *TreeNode) bool {
	stack := []*TreeNode{}
	pre := math.MinInt64
	for len(stack) > 0 || root != nil {
		// 先到最左下角
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}
		root = stack[len(stack)-1]   //栈顶元素
		stack = stack[:len(stack)-1] // 出栈
		if root.Val <= pre {
			return false
		}
		pre = root.Val
		root = root.Right
	}
	return true

}

File: 0098_validate-binary-search-tree/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        # 递归法， 时间，空间复杂度都为O(n)
        def dfs(node: TreeNode, left, right) -> bool:
            if node is None:
                return True
            if left < node.val < right:
                return dfs(node.left, left, node.val) and dfs(node.right, node.val, right)
            return False

        return dfs(root, -float('inf'), float('inf'))
